File: test/benchmark/larceny/deriv.js
// Larceny Benchmark: deriv (Node.js)
// Symbolic differentiation of expression trees
'use strict';

function deriv(e) {
    if (e.t === 0) return { t: 0, v: 0 };
    if (e.t === 1) return { t: 0, v: 1 };
    if (e.t === 2) {
        return { t: 2, l: deriv(e.l), r: deriv(e.r) };
    }
    // t === 3: product rule d(a*b) = a*db + da*b
    const dl = deriv(e.l);
    const dr = deriv(e.r);
    return { t: 2, l: { t: 3, l: e.l, r: dr }, r: { t: 3, l: dl, r: e.r } };
}

function countNodes(e) {
    if (e.t === 0 || e.t === 1) return 1;
    return 1 + countNodes(e.l) + countNodes(e.r);
}

function makeExpr() {
    const c3 = { t: 0, v: 3 };
    const c2 = { t: 0, v: 2 };
    const c5 = { t: 0, v: 5 };
    const x = () => ({ t: 1, v: 0 });
    const m1 = { t: 3, l: c3, r: x() };
    const m2 = { t: 3, l: m1, r: x() };
    const m3 = { t: 3, l: m2, r: x() };
    const m4 = { t: 3, l: c2, r: x() };
    const m5 = { t: 3, l: m4, r: x() };
    const a1 = { t: 2, l: m3, r: m5 };
    const a2 = { t: 2, l: a1, r: x() };
    return { t: 2, l: a2, r: c5 };
}

function benchmark() {
    let result = 0;
    for (let iter = 0; iter < 5000; iter++) {
        const e = makeExpr();
        const d = deriv(e);
        result = countNodes(d);
    }
    return result;
}

function main() {
    const __t0 = process.hrtime.bigint();
    const result = benchmark();
    const __t1 = process.hrtime.bigint();
    if (result === 45) {
        process.stdout.write("deriv: PASS\n");
    } else {
        process.stdout.write("deriv: FAIL result=" + result + "\n");
    }
    process.stdout.write("__TIMING__:" + Number(__t1 - __t0) / 1e6 + "\n");
}

main();

File: test/benchmark/larceny/python/deriv.py
#!/usr/bin/env python3
# Larceny Benchmark: deriv (Python)
# Symbolic differentiation of expression trees
import time

# Node types: 0=const, 1=var, 2=add, 3=mul


def make_const(v):
    return (0, v, None, None)


def make_var():
    return (1, 0, None, None)


def make_add(l, r):
    return (2, 0, l, r)


def make_mul(l, r):
    return (3, 0, l, r)


def deriv(e):
    t = e[0]
    if t == 0:
        return make_const(0)
    if t == 1:
        return make_const(1)
    if t == 2:
        return make_add(deriv(e[2]), deriv(e[3]))
    # t == 3: product rule d(a*b) = a*db + da*b
    dl = deriv(e[2])
    dr = deriv(e[3])
    return make_add(make_mul(e[2], dr), make_mul(dl, e[3]))


def count_nodes(e):
    t = e[0]
    if t == 0 or t == 1:
        return 1
    return 1 + count_nodes(e[2]) + count_nodes(e[3])


def make_expr():
    c3 = make_const(3)
    c2 = make_const(2)
    c5 = make_const(5)
    x = make_var
    m1 = make_mul(c3, x())
    m2 = make_mul(m1, x())
    m3 = make_mul(m2, x())
    m4 = make_mul(c2, x())
    m5 = make_mul(m4, x())
    a1 = make_add(m3, m5)
    a2 = make_add(a1, x())
    return make_add(a2, c5)


def benchmark():
    result = 0
    for _ in range(5000):
        e = make_expr()
        d = deriv(e)
        result = count_nodes(d)
    return result


def main():
    t0 = time.perf_counter_ns()
    result = benchmark()
    t1 = time.perf_counter_ns()

    if result == 45:
        print("deriv: PASS")
    else:
        print(f"deriv: FAIL result={result}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()
